0022-circuit-block.md

```diff
@@ -0,0 +1,185 @@
+            label=label,
+        )
+```
+
+Being ``Instruction``s, they can be appended to ``QuantumCircuit``s simply by specifying the lists of ``qargs`` and ``clargs``, for example via a dedicated ``block`` function wrapping the ``append`` method of ``QuantumCircuit``:
+```
+class QuantumCircuit:
+    ...
+
+    def block(
+        self, 
+        circuit: QuantumCircuit,
+        twirling_strategy: ...,
+        qargs: Optional[list[QubitSpecifier]] = None,
+        cargs: Sequence[ClbitSpecifier] | None = None,
+    ) -> InstructionSet:
+        from .block import CircuitBlock
+
+        # These use a `dict` not a `set` to guarantee a deterministic order to the arguments.
+        qargs = tuple(
+            {q: None for qarg in qargs for q in self._qbit_argument_conversion(qarg)}
+        )
+        cargs = tuple(
+            {b: None for carg in cargs for b in self._clbit_argument_conversion(carg)}
+        )
+
+        return self.append(CircuitBlock(circuit, twirling_strategy), qargs, cargs)
+```
+
+The logic in the ``block`` function above returns an ``InstructionSet`` containing a single ``CircuitInstruction``, with the block as ``operation`` and the correct register's slices as ``qubits`` and ``clbits``:
+```
+block = QuantumCircuit(2)
+block.cx(1, 0)
+
+qreg1 = QuantumRegister(3, "qreg1")
+qreg2 = QuantumRegister(3, "qreg2")
+
+circuit = QuantumCircuit(qreg1, qreg2)
+circuit.s(qreg1[1])
+circuit.h(qreg2[1])
+my_block = circuit.block(block, [1, 5])
+# equivalent to: my_block = circuit.block(block, [qreg1[1], qreg2[2]])
+
+print(my_block[0])
+>>> "CircuitInstruction(operation=Instruction(name='block', num_qubits=2, num_clbits=0, params=[]), qubits=(Qubit(QuantumRegister(3, 'qreg1'), 1), Qubit(QuantumRegister(3, 'qreg2'), 2)), clbits=())"
+
+print(circuit.draw())
+>>> qreg1_0: ────────────────
+>>>          ┌───┐┌────────┐
+>>> qreg1_1: ┤ S ├┤0       ├─
+>>>          └───┘│        │
+>>> qreg1_2: ─────┤        ├─
+>>>               │        │
+>>> qreg2_0: ─────┤  Block ├─
+>>>          ┌───┐│        │
+>>> qreg2_1: ┤ H ├┤        ├─
+>>>          └───┘│        │
+>>> qreg2_2: ─────┤1       ├─
+>>>               └────────┘
+```
+
+Similar to `for_loop` and `switch`, blocks can also be appended within a context:
+```
+with circuit.block(context=my_context, name="foo") as block:
+    block.h(0) 
+    block.cx(0, 1) 
+    block.cx(5, 6) # note that for_loop, switch and other ControlFlowOps have magic methods that
+                   # lets these indices, (5,6) here, be wrt to `circuit`, converting them to
+                   # appropriate input to Block
+```
+
+## Outstanding questions:
+1. In the existing implementation of blocks that exists in `pec-runtime`, blocks are hashable, so that they can be used as keys in dictionaries (specifically in the dictionary porduced by the ``NoiseLearner`` that maps blocks to noise models). This is clearly unsafe since circuits are mutable. Can we find a way around this? 
```

Comment:
In particular, how are users going to tie noise models to blocks if we can't have a `dict[CircuitInstruction[Block], NoiseModel]` or equivalent. This is a crucial feature, so we need to have this planned for.